src/compiler/checker.ts

```diff
@@ -10363,22 +10363,29 @@ namespace ts {
             }
 
             if (isConstEnumObjectType(type)) {
-                // enum object type for const enums are only permitted in:
-                // - 'left' in property access
-                // - 'object' in indexed access
-                // - target in rhs of import statement
-                let ok =
-                    (node.parent.kind === SyntaxKind.PropertyAccessExpression && (<PropertyAccessExpression>node.parent).expression === node) ||
-                    (node.parent.kind === SyntaxKind.ElementAccessExpression && (<ElementAccessExpression>node.parent).expression === node) ||
-                    ((node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.QualifiedName) && isInRightSideOfImportOrExportAssignment(<Identifier>node));
-
-                if (!ok) {
-                    error(node, Diagnostics.const_enums_can_only_be_used_in_property_or_index_access_expressions_or_the_right_hand_side_of_an_import_declaration_or_export_assignment);
-                }
+                checkConstEnumPosition(node);
             }
             return type;
         }
 
+        function checkConstEnumPosition(node: Node) {

```

Comment:
Readability? Like 7 lines of code, it fits in my memory:
http://www.simplypsychology.org/short-term-memory.html

As for performance, JS compilers would inline the function.